NEW FormField Title tooltip

emteknetnz opened the pull request:
Issue https://github.com/silverstripe/silverstripe-asset-admin/issues/1214

maxime-rainville commented on client/src/styles/legacy/_forms.scss:
```diff
@@ -150,6 +150,23 @@ fieldset {
     }
   }
+
+  .title-tip.tip {
```
This should be in a SCSS file in the same folder as Tip.

emteknetnz replied:
Moved to its own scss file

maxime-rainville reviewed (changes requested):
Need to clarify if Label tip and Field tip can be use simultaneously. If not, TextField should suppress the label tip somehow.

My instinct would be to convert Tip to be a functional components, but that's not a hill I'm desperate to die on.

emteknetnz commented on client/src/components/Tip/Tip.js:
```diff
@@ -55,18 +60,37 @@ class Tip extends Component {
+    };
+
+    if (this.props.type === TIP_TYPES.BUTTON) {
+      const className = `tip font-icon-${icon} text-${iconColor} ${this.props.extraClass || ''}`.trim();
```
Updated

emteknetnz commented on client/src/components/Tip/Tip.js:
```diff
@@ -55,18 +60,37 @@ class Tip extends Component {
+        tabIndex: 0,
+        className
+      };
+      el = <span {...props}><i className={iconClassName} /></span>;
```
Updated both the tip types to use the Silverstripe Button component

emteknetnz commented on client/src/components/Tip/Tip.js:
```diff
@@ -55,18 +60,37 @@ class Tip extends Component {
+      'aria-expanded': open
+    };
+
+    if (this.props.type === TIP_TYPES.BUTTON) {
```
Now they're both using the Silverstripe Button component, the only difference between them is the classnames they use, so I've kept them in the same file

emteknetnz commented on client/src/components/FieldHolder/FieldHolder.js:
```diff
@@ -160,13 +144,50 @@ function fieldHolder(Field) {
     }
 
+    /**
+     * @returns {object}
```
Updated

emteknetnz commented on client/src/components/FieldHolder/FieldHolder.js:
```diff
@@ -160,13 +144,50 @@ function fieldHolder(Field) {
+    /**
+     * Build description
+     *
+     * @returns {object}
```
Updated

emteknetnz commented on client/src/components/Tip/Tip.js:
```diff
@@ -92,13 +116,16 @@ export const tipShape = {
 
 Tip.propTypes = {
   ...tipShape,
+  type: PropTypes.oneOf(Object.values(TIP_TYPES)),
```
Updated

maxime-rainville reviewed (commented):
@clarkepaul Seems like maybe we should only have one of those tool tip open at a time. Not sure what would happen if you ended up in a situation where one tool tip overlap with another. Also, there's no mouse or focus styles for the `(?)` icon.

![image](https://user-images.githubusercontent.com/1168676/127595772-e9df9919-5302-47a3-bf16-4b20debf7776.png)

maxime-rainville commented on client/src/components/Tip/Tip.scss:
```diff
@@ -0,0 +1,20 @@
+    &:active:hover,
+    // click and hold and move mouse off element
+    &:active:not(:hover) {
+      background: transparent;
```
You could probably avoid this by conditionally changing the button `color` prop to `link`.

emteknetnz replied:
Changing the color prop on the `<Button>` component doesn't fix this

maxime-rainville commented on client/src/components/Tip/tests/Tip-story.js:
```diff
@@ -44,8 +45,9 @@ storiesOf('Admin/Tip', module)
-              icon={selectV2('Icon (examples)', ['lamp', 'attention', 'flag'], 'lamp')}
+              icon={selectV2('Icon (examples)', ['lamp', 'menu-help', 'attention', 'flag'], 'lamp')}
               importance={selectV2('Importance', importanceLevels, 'normal')}
+              type={selectV2('Type', TIP_TYPES, TIP_TYPES.TEXT_FIELD)}
```
Given how this story is structured, having this prop here doesn't provide any value. 

We should have a tip on the label and one on the field. Maybe we need a different story for each use case. Or maybe this should be on a story specific to form fields.

emteknetnz replied:
I've split the two different types to be different things in the pattern-lib

emteknetnz commented on client/src/components/Tip/Tip.js:
```diff
@@ -1,7 +1,18 @@
+ * List of values that can be passed in the `type` prop of the `<Tip />` component.
+ * @type {{TITLE: string, TEXT_FIELD: string}}
+ */
+export const TIP_TYPES = {
```
The pre-existing name is "importance level", so I think "type" isn't confusing and makes the most sense

emteknetnz replied:
I've changed the tip importance level from 'type' to 'description'

emteknetnz commented on client/src/components/Tip/Tip.js:
```diff
@@ -1,7 +1,18 @@
+ * @type {{TITLE: string, TEXT_FIELD: string}}
+ */
+export const TIP_TYPES = {
+  TITLE: 'title',
```
I think "title" makes more sense than say label because the the tip is next to the form field title, rather than the `<label>` html tag, which to me seems less obvious since it requires more knowledge of the implementation

If we change the name here we'd want to be consistent all the way back to `FormField::setLabelTip()`, which doesn't seems less obvious to me what it does

emteknetnz commented on client/src/components/Tip/Tip.js:
```diff
@@ -1,7 +1,18 @@
+ */
+export const TIP_TYPES = {
+  TITLE: 'title',
+  TEXT_FIELD: 'text-field',
```
Updated

emteknetnz commented on client/src/components/Tip/Tip.js:
```diff
@@ -55,28 +66,40 @@ class Tip extends Component {
+      >
+        <PopoverBody aria-live="assertive" aria-relevant="additions">{content}</PopoverBody>
+      </Popover>
     ];
```
I've refactored quite a lot of Tip.js

emteknetnz commented:
@clarkepaul I've refactored the Tip so that it uses the 'legacy' Popover trigger - see https://reactstrap.github.io/components/popovers/ - what this means is that when you click outside of a popover it will close the old one.  This means that only one popover at a time can be open.

I've added focus/actives states where I've changed the icon color because it's fairly easy to do.  Borders and box shadows are difficult since it's a round icon - the border/box-shadow will display as a square

I've added focus state when tabbing with a keyboard and mouse hover. 

It uses the `$input-focus-border-color` which is used for the text field border color when focusing in a field

![image](https://user-images.githubusercontent.com/4809037/127728117-b32a6ad1-32fc-4460-bb04-438198baa4d0.png)

I've used `darken($input-focus-border-color, 15%);` for the active state when clicking the icon

![image](https://user-images.githubusercontent.com/4809037/127728155-ed95ac9d-5445-480b-adb0-96cdd46f57ac.png)

Let me know if you'd like these colors changed

emteknetnz commented on client/src/components/Tip/Tip.js:
```diff
@@ -55,28 +66,40 @@ class Tip extends Component {
       fieldTitle
     });
 
+    let props = {
```
Updated

clarkepaul commented:
@emteknetnz @maxime-rainville nice that you got the popovers to close, that was put into the too hard basket a while back by @Cheddam so I'd think that there is an issue floating around for it.

@emteknetnz I'll get back to you on the colours of focus, active soon.

clarkepaul commented:
@emteknetnz I can't think of any other patterns we've used which would be any good apart from colour so I think you've gone in the right direction. Can you use the link colour (should output at #0071c4) for focus and active, use link:hover colour (should output at #0062ab) for just when it is hovered, they should contrast better.

emteknetnz commented:
@clarkepaul 
![image](https://user-images.githubusercontent.com/4809037/127796485-20076225-474a-43de-b315-fa7e1d0a01c9.png)
![image](https://user-images.githubusercontent.com/4809037/127796504-06afb5a8-4f27-4414-804c-ec95ac12cd5c.png)
![image](https://user-images.githubusercontent.com/4809037/127796520-dd80334e-0efd-465c-9d87-eb3d8a279009.png)

maxime-rainville reviewed (changes requested):
@clarkepaul Here's a video that shows the tooltip after and before: https://youtu.be/XbfIa7kdN1s 

One thing to note is that the tool tip will now be dismissed once you click somewhere else on the screen. I'm calling this an improvement over the status quo.

Tested in IE 11 and it worked great there as well.

@emteknetnz Massive improvement! I'm still not happy with having two things reffered as `type` even if it doesn't affect the API we expose to the rest of the system. Rename one of those things. But otherwise it looks good to merge pending Paul's approval.

maxime-rainville commented on client/src/components/Tip/Tip.js:
```diff
@@ -27,58 +39,40 @@ const tipImportanceMap = {
+  const buttonId = `${id}-tip`;
+  const buttonProps = {
+    id: buttonId,
+    onClick: () => {},
```
What's the point of onClick here?

emteknetnz replied:
The button component (presumably the reactstrap one) wants an onClick function defined otherwise it doesn't want to work.  So I put a noop in

emteknetnz commented on client/src/components/Tip/Tip.js:
```diff
@@ -27,58 +39,40 @@ const tipImportanceMap = {
 /**
- * UI element displaying a toggle-able "Tip". Designed to be used as an 'input group suffix'.
+ * UI element displaying a toggle-able "Tip".
+ * THe LABEL type will show next to a form field title label
```
Updated

maxime-rainville reviewed (changes requested):
Just two last tweaks and we should be good

emteknetnz commented:
@maxime-rainville done

maxime-rainville commented on client/src/components/Tip/Tip.js:
```diff
@@ -18,67 +30,49 @@ export const TIP_IMPORTANCE_LEVELS = {
+  const { content, fieldTitle, icon, id, importance } = props;
+  const { iconColor, description } = tipImportanceMap[importance];
+  const label = i18n.inject(i18n._t('Admin.TIP_LABEL', '{type} for {fieldTitle}'), {
+    description,
```
@emteknetnz Doesn't look like this one was done. Translation won't work if it's not actioned.

maxime-rainville reviewed (approved):
Looks beautiful!
